PDP: Working with data and schema

zachtil opened the pull request:
## Describe this PR

Add 4 new pages related to working with the Data Browser, Query Builder, and Schema tabs in PDP.

## Changes

* Several renames of files and folders and moving of files are necessary to support a navigation with more entries (basically, high impact on file and folder numbering)
* New content
    * Data Browser
    * Query Builder
    * Schema
    * Team members
    * Troubleshooting

martzoukos reviewed (changes requested):
It's mostly looking great! 

One thing to follow along about the DataBrowser is that we're currently changing the way filters work (tickets [CP-2160](https://linear.app/prisma-company/issue/CP-2160/implement-always-visible-filter-state) and [CP-2162](https://linear.app/prisma-company/issue/CP-2162/implement-composite-filter-state)). Once they get deployed, sometime next week probably, the filter bit of the DB page will need to be updated a bit.

martzoukos commented on content/800-data-platform/04-working-with-data/400-schema.mdx:
```diff
@@ -0,0 +1,27 @@
+
+You can view the Prisma schema of the current project environment in the **Schema** tab.
+
+Under the Schema title, you can view when your Prisma schema for the current environment was last updated indicated by the `Last pulled from GitHub 1 hour ago` line.
```
It might be worth mentioning somewhere here that the Schema file you're viewing is pulled from the specific Github branch that you added to your current environment. A common problem our users are facing is that they are telling us that they're not seeing their schema changes deployed and that's because they push to the wrong branch so it might be worth to also add this to the list of problems in the page.

We plan to improve that by showing the github branch in the page ([ticket](https://linear.app/prisma-company/issue/CP-2327/show-gh-branch-in-the-schema-page)) but until then, the Docs might help some people out.

martzoukos replied:
(that's the only change I'm suggesting, btw, sorry for the unorthodox double PR review 😅 )

zachtil replied:
> 

I've added this to the Schema page.

If you select another environment which uses a separate branch or repository, the Schema tab shows the schema file from that branch or repository.

zachtil commented:
> It's mostly looking great!
> 
> One thing to follow along about the DataBrowser is that we're currently changing the way filters work (tickets [CP-2160](https://linear.app/prisma-company/issue/CP-2160/implement-always-visible-filter-state) and [CP-2162](https://linear.app/prisma-company/issue/CP-2162/implement-composite-filter-state)). Once they get deployed, sometime next week probably, the filter bit of the DB page will need to be updated a bit.

@martzoukos - one of the tickets has now been dropped? and the other one is still in progress. I'll address the changes when back on Jul 25.

janpio commented on content/200-concepts/100-components/06-prisma-studio.mdx:
```diff
@@ -11,6 +11,12 @@ Prisma Studio is a visual editor for the data in your database. Note that Prisma
+<Admonition>
+
+**Note**<br /><br />
+The features of Prisma Studio are identical to those in the Data Browser in the Prisma Data Platform. See the [Data Browser](/data-platform/data-browser/) documentation.
```
```suggestion
The features of Prisma Studio are similar to those in the Data Browser in the Prisma Data Platform. See the [Data Browser](/data-platform/data-browser/) documentation.
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+
+</TopBlock>
+
+## Models (Tables)
```
On MongoDB these are collections:
```suggestion
## Models (Tables or Collections)
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+
+</TopBlock>
+
+## Models (Tables or Collections)
+
+When you first open the **Data Browser**, you can review the tables in your application database which are based on the models defined in the Prisma schema file.
+
+<img src="https://res.cloudinary.com/prismaio/image/upload/v1657146172/docs/pdp/data-browser-models-view_zs1ihu.png" alt="Data Browser - Models view"/>
+
+You can select a table and its data opens in a new tab.
```
This is a comment that applies to all the content here:
In the past we tried to use Prisma terms when talking about Prisma, and only fall back to the database terms when talking explicitly about the database. Here for example one could have a headline "Models" and then just talk about how you can see all your models and their data.

tanberry replied:
Also need a comma after "database", before "which". :-)

zachtil replied:
Thanks for the feedback, Jan. I'll fix all of these.

zachtil replied:
So I did. And added a thing of my own.

I've added 3 admonitions defining `model`, `field`, `record` right at the beginning of those sections and connected those to their relational db and MongoDB counterparts. Also added links to Prisma ORM docs to cross-link to the context where these are defined.

I am obviously new to Prisma still and if it makes sense to cater to this need, we could leave them there. Or redo them in a more meaningful way.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+
+You can select a table and its data opens in a new tab.
+
+### Open and close tables
+
+To open another table, click the **+** button.
+
+To close a table, click the the **x** button in the table tab.
+
+<img src="https://res.cloudinary.com/prismaio/image/upload/v1657146172/docs/pdp/data-browser-close-open-table_g7xho7.png" width="375px" alt="Data Browser - Open and close tables"/>
```
```suggestion
### Open and close models

To open another model, click the **+** button.

To close a model, click the the **x** button in the model tab.

<img src="https://res.cloudinary.com/prismaio/image/upload/v1657146172/docs/pdp/data-browser-close-open-table_g7xho7.png" width="375px" alt="Data Browser - Open and close models"/>
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+
+### Filter data
+
+Use the **Filters** menu to filter data in the table by adding conditions.
```
```suggestion
Use the **Filters** menu to filter data in the model by adding conditions.
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+      - **gte**
+      - **not**
+   3. Type the value you want to use for the filter.<br />
+      **Step result**: The table is filtered immediately after you define the first filter.
```
```suggestion
      **Step result**: The model is filtered immediately after you define the first filter.
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+
+**Result**
+
+* The data in the table is filtered based on the conditions you added.
```
```suggestion
* The data in the model is filtered based on the conditions you added.
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+* The data in the table is filtered based on the conditions you added.
+* The default **None** value in the **Filters** menu changes to the number of filters you added.
+
+### Show and hide columns
+
+You can select which columns to view or hide by using the **Fields** menu.
```
Is this about hiding columns or not including fields in the result?
(I again would use the Prisma term "fields")

zachtil replied:
Changed. Thanks for this!

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,142 @@
+
+Also, the **Fields** menu shows the number of fields that are currently selected.
+
+### Show and hide rows
+
+You can also select to show or skip a specific number of rows in the table view.
```
I think we usually call "rows" in Prisma context "entries".

zachtil replied:
We use "records" quite a bit more in the docs.

zachtil replied:
I've found only 6-7 occurrences of "entries" and they felt interchangeable with "records". Or am I missing something?

janpio replied:
No, I just don't like "records" that much so keep ignoring our usage 😆

zachtil replied:
Can you take our usage into account for the purposes of this ticket? :) And have a terminology discussion later?

zachtil replied:
@janpio I'll stick to using "records" here and will start a terminology discussion separately.

zachtil replied:
Update: I've started a docs issue to evaluate the switch to using 'entries' here: https://github.com/prisma/docs/issues/3533

It sounds like the widespread use of 'records' is due to the fact that MongoDB was only added this year and we didn't have a good umbrella term before that.

zachtil replied:
Resolving the convo here.

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+## Default model queries
+
+- `findMany()`
+- `findMany()` in which the `post_id` is 0
```
```suggestion
- `findMany()` in which the `id` is `example`
```
I think this is the actual default query

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+
+## Default model queries
+
+- `findMany()`
+- `findMany()` in which the `post_id` is 0
+- `findMany()` but return the data from only 2 fields
```
Here are what I think are the default queries with their current comment:

```
// Find the first 100 rows in the `User` table
await prisma.user.findMany({ take: 100 })

// Find the first 100 rows in the `User` table that match a criteria
await prisma.user.findMany({
  where: {
    "id": "example"
  },
  take: 100 
})

// Find the first 100 rows in the `User` table, but only return a few columns
await prisma.user.findMany({
  select: {
    "id": true,
    "createdAt": true
  },
  take: 100
})
```

zachtil replied:
I've replaced these with excerpts from the code sample. Thanks for pasting it here!

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+
+## Run a query
+
+You can test and run queries for your Prisma app from the Query Console and view their results on the right side of the screen.
```
"Prisma app" is not really a thing. "app using Prisma Client" maybe.

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+
+## Add a query
+
+The Query Console supports the autocompletion you get when you use the Prisma Client API in Visual Studio Code.
```
```suggestion
The Query Console supports the autocompletion you get when you use your Prisma Client API in Visual Studio Code.
```

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+**Result**
+
+* After you define a query, the green button is available for testing the query.
+* The Prisma Data Platform runs on a best-guess basis even if the query code is not completely defined. For example, in the following snippet, the `select` statement is incomplete, but the Prisma Data Platform still runs the `findMany()` query.
```
How does it do that?

zachtil replied:
That's what I found while testing.

zachtil replied:
Tested with a couple of very badly formed queries. `findMany` still runs.

janpio replied:
Not sure if bug or feature honestly. Sounds a bit scary. 
(For example with a `deleteMany` where the `WHERE` is misspelled - and then it deletes _everything_)

zachtil replied:
FYI I have removed the information about the best-effort because it seems to open a can of worms. I can address it later. Tracking in this docs issue:
https://github.com/prisma/docs/issues/3533

zachtil commented on content/800-data-platform/10-team-members/050-user-roles.mdx:
```diff
@@ -0,0 +1,43 @@
+| Edit data | ✅ | ✅ | ✅ | ✅ | ⛔️ |
+| **Query Console** |  |  |  |  |  |
+| Run default queries | ✅ | ✅ | ✅ | ✅ | ✅ |
+| Add new queries | ✅ | ✅ | ✅ | ✅ | ✅ |
```
No, I had an incorrect assumption that we might be preventing users with specific roles to write their own queries in case of us not willing to allow them to overload the database or something to that end.

I've now merged "run and add" into a single row.

janpio replied:
"Add" indicated that they would still be there for other users that also can access the account.
I am not sure if that is the case, or they are only stored for myself or even not at all.

zachtil replied:
Wow, so we completely came thinking about this from very different directions.

I added a query as a 'Viewer' and its stored there for my personal later usage.

I added another one as the 'Project owner' and it's the same. Neither can see what the other added.

zachtil replied:
It still makes sense to me that 'Add and run queries' is available for all user roles.

But you could have a different view on this so I'd be glad to hear.

zachtil commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+2. Run the query using one of the 3 available method.
+   - Click the **Run active query** button in the upper-left.
+   - Click the **Play** (**Run query**) button at the left margin of the editor.
+   - Press Cmd + Enter (on macOS) or Ctrl + Enter (on Windows).
```
I wasn't aware of that. So cool!

zachtil commented:
Thanks for the review and for pushing the correct terminology here @janpio . I've fixed what I think to be all incorrect instances. And as mentioned above, added admonitions for my own needs which I think might be useful for novices jumping into Prisma through the Prisma Data Platform.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+
+## Models (tables or collections)
+
+When you first open the **Data Browser**, you can review the models in your application database, which are based on the models defined in the Prisma schema file.
```
Maybe simpler:
```suggestion
When you first open the **Data Browser**, you see a list of all the models defined in your Prisma schema file.
```

zachtil replied:
Thanks, Jan. Changed.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+
+**What is a record?**<br /><br />
+
+A **record** refers to a row of data in a table in a relational database (PostgreSQL, MySQL, SQL Server, SQLite, CockroachDB) or a **document** in MongoDB.
```
```suggestion
A **record** refers to a **row of data in a table** in a relational database (PostgreSQL, MySQL, SQL Server, SQLite, CockroachDB) or a **document** in MongoDB.
```

zachtil replied:
Thanks! Changed.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+
+In the model view, you can edit data directly in the model cells. You can add new records as well as delete existing records.
+
+You must confirm every edit operation (add, edit, delete). You must confirm any new records that you add or edits of existing records that you make with the **Save change** button. 
```
This line sounds a bit inconsistent - what about deleting? WHy is that only in the brackets and not explained again later?

zachtil replied:
Good catch.

zachtil replied:
I've added more explanation around that.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+1. In the model view, click **Add record**.
+2. Based on the data allowed in each field, type the data for the record.
+
+   | Field data type | Description                                                                                                                                                                                                                                                         |
+   | :--------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+   |       `A?`       | Text                                                                                                                                                                                                                                                                |
+   |       `#`        | Integer<br /><br />If such a field has `autoincrement()` pre-filled, do not edit the cell and add a number.                                                                                                                                                        |
+   |       `[]`       | List<br /><br />It typically referes to a list of records that exist in another model in the database.                                                                                                                                                                  |
+   |       `{}`       | Relation with a model defined separately in the database.<br /><br />Typically, you need to select the same value as any of the previous records<br />Click the name of the model to see the list of values which you can then select for the related field. |
```
I think there might be more of these. DevProd team should be able to tell you.

janpio replied:
One for example is `Json`, and embedded documented (for MongoDB) [which shows the same as `Json` right now].

janpio replied:
This list could also be already listed in the "show data" section at the beginning - it describes what type a field/column is.

zachtil replied:
I've added information about the `Json` field.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+The model is immediately filtered to show or hide records based on your selection.
+
+The **Showing** menu indicates how many records are shown out of how many available records are in the model.
+
```
I think you could also document that you can sort fields/columns by clicking on them.

zachtil replied:
Added.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+
+### Edit a record
+
+1. Double-click in a cell with existing date to edit its contents.
```
```suggestion
1. Double-click in a cell with existing data to edit its contents.
```

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+
+You can add multiple records and edited cells and, thus, accumulate them.
+
+In the end, click **Save&nbsp;<number\>&nbsp;changes** to finalize them.
```
`<number\>` is unexpected notation.

zachtil replied:
Removed this. Yes, it doesn't make sense.

janpio commented on content/800-data-platform/05-data-browser.mdx:
```diff
@@ -0,0 +1,165 @@
+<Admonition type="warning">
+
+**Warning**<br /><br />
+Deleting a record is a separate operation that force-saves any unsaved edits and is then completed.
```
Bug or feature?

zachtil replied:
Not sure. It is the current behavior. I've edited the text to hopefully make it clearer. I'll track this conversation with Spiros to figure out if it's a bug.

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+
+| UI component          | Description                                                                         |
+| --------------------- | ----------------------------------------------------------------------------------- |
+| Query Editor          | • Add, edit, delete queries using the Prisma Client API<br />• Run a selected query |
```
No unordered Markdown lists in tables? 💩

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+
+## Default model queries
+
+- `findMany()` for the first 100 records in the `User` model
+- `findMany()` for the first 100 records in the `User` model that match a criteria
+- `findMany()` for the first 100 records in the `User` model but only return data from 2 fields
```
Is this always the `User` query or not just the first one or something like that? DevProd team can probably tell you.

zachtil replied:
Added a note that this is from a project that specifically has the `User` model.

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+**Steps**
+
+1. In the Query Editor, place the cursor anywhere in a query block.
+2. Run the query using one of the 3 available method.
```
```suggestion
2. Run the query using one of the 3 available methods.
```

zachtil replied:
Good catch. Changed.

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+2. Run the query using one of the 3 available method.
+   - Click the **Run active query** button in the upper-left.
+   - Click the **Play** (**Run query**) button at the left margin of the editor.
+   - Press <kbd>Cmd</kbd> + <kbd>Enter</kbd> (on macOS) or <kbd>Ctrl</kbd> + <kbd>Enter</kbd> (on Windows).
```
Beautiful:
<img width="356" alt="image" src="https://user-images.githubusercontent.com/183673/178057609-4c5633e6-fa3b-483f-bb13-9b1f5d82caea.png">

janpio commented on content/800-data-platform/06-query-console.mdx:
```diff
@@ -0,0 +1,68 @@
+
+## Add a query
+
+The Query Console supports the autocompletion you get when you use your Prisma Client API in Visual Studio Code.
```
```suggestion
The Query Console supports the same autocompletion you get when you use your Prisma Client API in Visual Studio Code or other modern IDEs.
```

zachtil replied:
Changed as you suggested. Also added 'syntax highlighting' as that is also available.

janpio reviewed (approved):
So much better!

A few nit comments and fixes mostly, but very close to ready I would say.
